gitbook-docs/data_model.md

```diff
@@ -264,6 +264,64 @@ the value should be merged to the full model mounted where the delta‘s context
 }
 ```
 [<]: #
+
+## History format
+
+The history format is available since SignalK 1.1 and is designed to efficiently
+represent and transfer information over a period of time.
+
+It can be used as a response from a server to a client (for example over HTTP)
+or as a file format, in which case the recommended extension is `.skl`.
+
+[>]: # (mdpInsert ```json fsnip ../samples/history/basic.json)
+```json
+{
+  "version": "1.1.0",
+  "startDate": "2018-10-06T04:00:00Z",
+  "endDate": "2018-10-06T04:00:02Z",
+  "generator": "Hand-written with love with the spec",
+
+  "objects": [
+    {
+      "context": "vessels.urn:mrn:xxx",
+      "timestamps": [ "2018-10-06T04:00:00Z", "2018-10-06T04:01:00Z", "2018-10-06T04:00:02Z"],
```

Comment:
Is the history always contiguous? If yes the timestamps as JSON strings are redundant. Parsing them to any native datetime format takes also a nontrivial amount of effort, compared to generating time data during processing.

Should there be two formats, one for arbitrary timeseries with gaps and another for `(start, end, periodlength)` data?

Reply:
I think having a history-like format for data without requiring that the history be contiguous is helpful. Making the history contiguous and regularly spaced is very hard. Making sure that all objects in the file have data on the same rhythm is also hard.

I think having two formats (or two variants of one format) might make sense here.

Reply:
One word of caution though: if start/end/period is defined at the top-level then all AIS target will need to provide values for every individual timestamp which means probably a huge number of `null` in the file.

If we think one main vessel and a bunch of AIS targets is going to be a frequent and important use-case (I do) then the start/end should be defined at the vessel level.

Reply:
>Making the history contiguous and regularly spaced is very hard. 

Totally depends on the implementation. For example InfluxDb does it for you out of the box - nulls for holes. 

Often you want data in coarse time intervals. Then some small holes or having different sampling intervals are not a problem, but missing head or tail parts naturally are.

Reply:
An important use case is "give me all history between these times", eg I want a copy of all the raw data. So assuming there will always be a timeslice is not valid. 
Also assuming startTime+(timeslice*n)= actual time is usually going to be valid, but will fail erratically when we have nanos in the db (influx) and ms in the data. Key values may move from one timeslice to another.

Reply:
For raw data retrieval I don’t think we really need anything else than deltas in an array, possibly with a bit of metadata in the header.

Reply:
But then you have two formats for history depending on the query. Pushing out gzipped deltas works for both cases, and  solves the RAM problems at both ends as they can be written and read/processed as a stream